framework/src/mfem/markers/MFEMRefinementMarker.C

```diff
@@ -0,0 +1,122 @@
+//* This file is part of the MOOSE framework
+//* https://mooseframework.inl.gov
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifdef MOOSE_MFEM_ENABLED
+
+#include "MFEMRefinementMarker.h"
+#include "MFEMIndicator.h"
+
+registerMooseObject("MooseApp", MFEMRefinementMarker);
+
+InputParameters
+MFEMRefinementMarker::validParams()
+{
+  InputParameters params = MFEMGeneralUserObject::validParams();
+  params.registerBase("Marker");
+
+  MooseEnum refinement_type("H_REF P_REF H_P_REF", "H_P_REF");
+  params.addRequiredParam<MooseEnum>(
+      "refinement_type",
+      refinement_type,
+      "Specifies whether to use h-refinement, p-refinement or both.");
+
+  params.addRequiredParam<Real>("refine", "Error fraction for the refiner");
+  params.addRangeCheckedParam<Real>("refine",
+                                    0,
+                                    "refine>=0 & refine<=1",
+                                    "Elements within this percentage of the max error will "
+                                    "be refined.  Must be between 0 and 1!");
+  params.addRangeCheckedParam<int>(
+      "max_h_level", 0, "max_h_level>=0 & max_h_level <= 10", "Total number of h-refinement steps");
+  params.addRangeCheckedParam<int>(
+      "max_p_level", 0, "max_p_level>=0 & max_p_level <= 10", "Total number of p-refinement steps");
+
+  params.addRequiredParam<std::string>("indicator", "Estimator to use");
+  return params;
+}
+
+MFEMRefinementMarker::MFEMRefinementMarker(const InputParameters & params)
+  : MFEMGeneralUserObject(params),
+    _estimator_name(getParam<std::string>("indicator")),
+    _error_threshold(getParam<Real>("refine")),
+    _max_h_level(getParam<int>("max_h_level")),
+    _max_p_level(getParam<int>("max_p_level")),
+    _refinement_type(getParam<MooseEnum>("refinement_type"))
+{
+  // _use_h_refinement and _use_p_refinement both default to false
+  if (_refinement_type == "H_P_REF")
+  {
+    _use_h_refinement = true;
+    _use_p_refinement = true;
+  }
+  else if (_refinement_type == "H_REF")
+  {
+    _use_h_refinement = true;
+  }
+  else if (_refinement_type == "P_REF")
+  {
+    _use_p_refinement = true;
+  }
+}
+
+void
+MFEMRefinementMarker::setUp()
+{
+  // fetch const ref to the estimator
+  const auto & estimator = getUserObjectByName<MFEMIndicator>(_estimator_name);
```

Comment:
We do this below as well, so might as well define an `_estimator` member perhaps?